### PR TITLE
lbfgs: support passing in `previous_optimizer_results`

### LINKED ISSUE #924: tfp.optimizer.lbfgs_minimize calls function for converged models
As far as I can tell, `lbfgs_minimize` continues to call the passed in `value_and_gradients`  function with the fullet set of parameters, even if some of the models in the batch have converged. When passing in a `tf.function` as `value_and_gradients`, I don't think this can be optimised away.

Would it be possible to have `value_and_gradients` only be called with the subset of the models that haven't converged yet?

We've previously implemented a batch optimizer that worked on a `(num_models, num_parameters)` tensor and call the underlying function on a slice of that tensor which included only the unconverged models.

Since unconverged models may not necessarily be contiguous, this worked by swapping out a newly converged model's parameters with the last one in the tensor and then decreased the slice by 1. As you do this, you can keep track of the permutation you've applied so far, and at the end unpermute the parameters.

I can see a few issues with meshing this with the current implementation of LBFGS. First, the `lbfgs_minimize` function accepts any number of batching dimensions, while slicing the way I described only works for a single batching dimension. Secondly, it would require the `value_and_gradients` function to accept a dynamic shape tensor (for the first dimension).

I don't know what the best way to implement this while keeping the curent API is. We could just have a boolean that controls this behaviour (and asserts the shape of the parameters is a matrix). 

Is this a good idea? Is there any desire for a PR implementing this?

**Edit** I submitted a PR #928 that proposes a minimal change to the API of `lbfgs_minimize` allowing optimization to be resumed after a single model converges. In between calls to `lbfgs_minimize` the user can do the swapping out of model as appropriate.

### PR DESCRIPTION
**Pull request by cristicbz**
This is a very small change that adds support for resuming lbfgs minimization if previously stopped by the stopping condition. If this design isn't the preferred one, happy to change the PR---I could also change the name of the argument to `experimental_initial_state` if you think it's not necessarily the way the API should look.

This means you can run the optimizer with `converged_any` and every time a single model converges from the batch, you can swap out the converged model and continue optimizing. For our problem, this gives us a 15-40% speed-up; we use batch optimisation for different hyperparameter settings, and strongly regularised models converge much quicker. Being able to stop evaluating them for the remaining iterations is an important optimisation.

Fixes #924 

Here's some pseudo-python ripped from our codebase as an example of how this can be used to speed up optimisation:

<details>

```python
num_models = ... # number of models simultaneously being optimized
full_parameters = tf.Variable(
  ...,
  shape=(num_models, num_model_parameters),
)
results = None

active_prefix = num_models
initial_parmaters = full_parameters.read_value()

total_permutation = np.arange(num_models)
while True:
    # Run LBFGS until one of the models converges.
    results = tfp.optimizer.lbfgs_minimize(
        compiled_loss,
        initial_parameters,
        stopping_condition=tfp.converged_any,
        max_iterations=max_iterations,
        initial_state=results,
    )
    # Set the initial parameters to `None` since in future runs, we'll
    # use the previous state instead .
    initial_parmeters = None

    # Update the `full_parameters` with their new positions
    full_parameters[:active_prefix].assign(results.position)

    # If we actually hit `max_iterations` or all models converged, then
    # break out of the loop.
    inactive = converged.numpy() | failed.numpy()
    if (
        results.num_iterations == max_iterations
        or inactive.all()
    ):
        # All models converged.
        break

    # Compute new permutation by swapping converged models with the
    # models at the end of the current `active_prefix`.
    new_permutation = np.arange(num_models)
    new_inactive_indices = list(inactive.nonzero()[0])
    new_prefix = active_prefix - len(new_inactive_indices)
    swap_with = list(range(new_prefix, active_prefix))
    new_permutation[
        new_inactive_indices + swap_with
    ] = new_permutation[swap_with + new_inactive_indices]

    # Apply new permutation to the parameters and optimizer state.
    full_parameters.assign(
        tf.gather(full_parameters.read_value(), new_permutation)
    )
    results = results._replace(
        converged=tf.zeros((new_prefix,), dtype=tf.bool),
        failed=tf.zeros((new_prefix,), dtype=tf.bool),
        position=tf.gather(
            results.position, new_permutation[:new_prefix], axis=0
        ),
        objective_value=tf.gather(
            results.objective_value,
            new_permutation[:new_prefix],
            axis=0,
        ),
        objective_gradient=tf.gather(
            results.objective_gradient,
            new_permutation[:new_prefix],
            axis=0,
        ),
        position_deltas=tf.gather(
            results.position_deltas,
            new_permutation[:new_prefix],
            axis=1,
        ),
        gradient_deltas=tf.gather(
            results.gradient_deltas,
            new_permutation[:new_prefix],
            axis=1,
        ),
    )

    # Keep track of overall permutation.
    total_permutation = total_permutation[new_permutation]
    active_prefix = new_prefix

inverse_total_permutation = np.arange(num_models)
for i_from, i_to in enumerate(total_permutation):
    inverse_total_permutation[i_to] = i_from

full_parameters.assign(
  tf.gather(parameters.read_value(), inverse_total_permutation),
)
```

</details>

### CODE REVIEW

**Comment by cristicbz** [comment]
@csuter Is there some way for me to get this PR looked at---I understand it may not be top priority, but I think it's a pretty small change that makes the lbfgs optimizer significantly more useful. Thanks!

File: tensorflow_probability/python/optimizer/lbfgs.py [hunk 1]
```diff
@@ -79,6 +79,7 @@
 
 def minimize(value_and_gradients_function,
              initial_position,
+             initial_state=None,
```
Comment (csuter) [inline tensorflow_probability/python/optimizer/lbfgs.py:82]: let's call this `previous_optimizer_results`, which mirrors similar functionality and naming style in the MCMC package (see `previous_kernel_results` argument to `sample_chain`).

Comment (cristicbz) [reply tensorflow_probability/python/optimizer/lbfgs.py]: Done, thanks!

File: tensorflow_probability/python/optimizer/lbfgs.py [hunk 2]
```diff
@@ -209,9 +216,23 @@ def quadratic_loss_and_gradient(x):
...
   with tf.name_scope(name or 'minimize'):
-    initial_position = tf.convert_to_tensor(
-        initial_position, name='initial_position')
-    dtype = initial_position.dtype.base_dtype
+    if initial_position is not None:
+      if initial_state is not None:
+        raise ValueError(
+            'If `initial_position` is not `None` then `initial_state` must be '
+            '`None`.'
+        )
```
[the remark this answers is not in the document]
Comment (cristicbz) [reply tensorflow_probability/python/optimizer/lbfgs.py:225]: Done.

File: tensorflow_probability/python/optimizer/lbfgs.py [hunk 3]
```diff
@@ -209,9 +216,23 @@ def quadratic_loss_and_gradient(x):
     stopping_condition = bfgs_utils.converged_all
 
   with tf.name_scope(name or 'minimize'):
-    initial_position = tf.convert_to_tensor(
-        initial_position, name='initial_position')
-    dtype = initial_position.dtype.base_dtype
+    if initial_position is not None:
```
[the remark this answers is not in the document]
Comment (cristicbz) [reply tensorflow_probability/python/optimizer/lbfgs.py:220]: I think you mean `==` rather than `!=` above, but yeah you're right it's cleaner, done.

**Comment by cristicbz** [comment]
Thanks for the review @csuter , I amended as per your comments, please take a look!

File: tensorflow_probability/python/optimizer/lbfgs.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (csuter) [reply tensorflow_probability/python/optimizer/lbfgs.py:225]: indeed -- thanks! :)

File: tensorflow_probability/python/optimizer/lbfgs.py [hunk 4]
```diff
@@ -143,6 +144,13 @@ def quadratic_loss_and_gradient(x):
     initial_position: Real `Tensor` of shape `[..., n]`. The starting point, or
       points when using batching dimensions, of the search procedure. At these
       points the function value and the gradient norm should be finite.
+      Exactly one of `initial_position` and `previous_optimizer_results` can be
+      non-None.
+    previous_optimizer_results: An `LBfgsOptimizerResults` namedtuple to
+      intialize the opitimizer state from, instead of an `initial_position`.
```
Comment (csuter) [inline tensorflow_probability/python/optimizer/lbfgs.py:150]: typo: s/opitimizer/optimizer/

Comment (cristicbz) [reply tensorflow_probability/python/optimizer/lbfgs.py]: oops, thanks!

**Comment by cristicbz** [comment]
@csuter Thanks for the approval! I fixed that last typo and tests passed, so I think this is good to go?

**Comment by csuter** [comment]
Thanks for the contribution @cristicbz, and thanks for your patience in the review process. Sorry we took so long! The change is checked in to our internal version control and should show up here momentarily.

**Pull request merged**